stablehlo/reference/Ops.cpp

```diff
@@ -527,6 +543,73 @@ Tensor evalDivideOp(const Tensor &lhs, const Tensor &rhs,
   return result;
 }
 
+Tensor evalDotGeneralOp(const Tensor &lhs, const Tensor &rhs,
+                        const Axes &lhsBatchingDimensions,
+                        const Axes &rhsBatchingDimensions,
+                        const Axes &lhsContractingDimensions,
+                        const Axes &rhsContractingDimensions,
+                        ShapedType resultType) {
+  Tensor result(resultType);
+  Axes lhsNonBatchingNonContractingDims;
+  for (auto i = 0; i < lhs.getType().getRank(); ++i)
+    if (!llvm::is_contained(lhsBatchingDimensions, i) &&
+        !llvm::is_contained(lhsContractingDimensions, i))
+      lhsNonBatchingNonContractingDims.push_back(i);
+
+  Axes rhsNonBatchingNonContractingDims;
+  for (auto i = 0; i < rhs.getType().getRank(); ++i)
+    if (!llvm::is_contained(rhsBatchingDimensions, i) &&
+        !llvm::is_contained(rhsContractingDimensions, i))
+      rhsNonBatchingNonContractingDims.push_back(i);
+
+  Sizes contractingDimSizes;
+  int64_t totalContractingSize = 1;
+  for (uint64_t i = 0; i < lhsContractingDimensions.size(); ++i) {
+    contractingDimSizes.push_back(
+        lhs.getType().getShape()[lhsContractingDimensions[i]]);
+    totalContractingSize *=
+        lhs.getType().getShape()[lhsContractingDimensions[i]];
+  }
+
+  for (auto resultIt = result.index_begin(); resultIt != result.index_end();
+       ++resultIt) {
+    Index lhsIdx(lhs.getType().getRank());
+    Index rhsIdx(rhs.getType().getRank());
+    int64_t resultDim = 0;
+
+    // Indices do not change for batching dimensions.
+    for (uint64_t i = 0; i < lhsBatchingDimensions.size(); ++i, ++resultDim) {
+      lhsIdx[lhsBatchingDimensions[i]] = (*resultIt)[resultDim];
+      rhsIdx[rhsBatchingDimensions[i]] = (*resultIt)[resultDim];
+    }
+
+    // The non-batching, non-contracting dimensions of the operands are copied
+    // over to the result dimensions following the batching dimensions.
+    for (uint64_t i = 0; i < lhsNonBatchingNonContractingDims.size(); i++)
+      lhsIdx[lhsNonBatchingNonContractingDims[i]] = (*resultIt)[resultDim++];
+
+    for (uint64_t i = 0; i < rhsNonBatchingNonContractingDims.size(); i++)
+      rhsIdx[rhsNonBatchingNonContractingDims[i]] = (*resultIt)[resultDim++];
+
+    auto sum = Element(resultType.getElementType(), 0.0);
+    // All pair-wise combination of contracting dimensions needs to be summed.
+    for (int64_t i = 0; i < totalContractingSize; ++i) {
+      sum = sum + lhs.get(lhsIdx) * rhs.get(rhsIdx);
+      if (contractingDimSizes.empty()) continue;
+      for (int64_t j = contractingDimSizes.size() - 1; j >= 0; --j) {
+        lhsIdx[lhsContractingDimensions[j]]++;
+        rhsIdx[rhsContractingDimensions[j]]++;
+        if (lhsIdx[lhsContractingDimensions[j]] != contractingDimSizes[j])
+          break;
+        lhsIdx[lhsContractingDimensions[j]] = 0;
+        rhsIdx[rhsContractingDimensions[j]] = 0;
+      }
+    }
+    result.set(*resultIt, sum);
+  }
+  return result;
+}
```

Comment:
I support diverging this implementation from the spec, which is way too complicated currently. Opened #1551 to track bringing the two into alignment, but I don't think this is a high priority.